Interface updates

mfitzp opened the pull request.

mfitzp committed:
Add PyCharm .idea folder to .gitignore

mfitzp committed:
Initial work on interface updated.

This is work-in-progress and currently looks quite a bit like
flask-wooey. It is also broken (nagivation, job submission).

Once all the views are completed (+working) the design will start
working towards the dashboard with sidebar elements, tiles, etc.

mfitzp committed:
Script output tile view

Script outputs are now shown on tile view (thumbnail-panels). The console
and file listings can be collapsed (or uncollapsed) as needed. The goal is
for the panels to support self-organising layout with dimensions/aspect ratio
to fit the content (e.g. 2x1 image will be given a different space to 1x1).

Chris7 commented:
Looks slick! Are you considering having the dashboard be a backend feature (meaning like a separate link) or have the main display be a dashboard?

mfitzp commented:
Not 100% sure yet actually.... I’m planning to create a user-dashboard for logged in users so they can get access to historical outputs, and ‘favourite’ outputs (e.g. plots they want to keep pinned in their dashboard for reference). There will also be a list of favourite scripts in there too to speed things up for people. For admin users we can include admin tools in this view (or as an alternate admin dash).

The script-listing will be dashboard-esque with tiles for each script (to allow e.g. images to illustrate what the script does) together with a search-filter interface with live filtering. Script outputs are also a bit dashboard-ish at the moment with tiles for each output - these will get tools e.g. for favouriting, and resizing/rearranging outputs.

Once the main views are done I’m going to add a config sidebar (slide-out) for accessing more administration functions.

I guess the only thing that isn’t dashboard-like is the job submission page (currently broken) which will show the form and the about /help text as a static page. But the idea will be to make it seem as seamless as possible between the views: a consistent ‘canvas’ that stuff is happening on. 
On 17 July 2015 at 15:01:22, Chris Mitchell (notifications@github.com) wrote:

Looks slick! Are you considering having the dashboard be a backend feature (meaning like a separate link) or have the main display be a dashboard?

—
Reply to this email directly or view it on GitHub.

mfitzp committed:
Job submission is now working

The template and views for submitted scripts to jobs is now working.
The view code is incomplete and a bit hacky, but works for testing.

mfitzp committed:
Add pinstripe background to highlight tile-widgets.

mfitzp committed:
Continuing development of the UI [migrations]

Ongoing work on the new UI. Note that this commit requires migrating
the database to add a new size_bytes column to files. This is the
start of adding support for file metadata support in the output lists.

mfitzp committed:
Refactoring of the wooey_celery task views JSON

This is a restructuring of the celery task views JSON to return
a global queue, user queue and user results set. Each data set is
available as an individual JSON (maybe unneccessary) and also as
a combined all_queues_json for the timer update.

The navbar header now auto-refreshes the status of the queues
as jobs are executed.

mfitzp committed:
Implementation of scrapbook.

This adds support for a Favorite model for bookmarking objects in the
system. Users can use these to favorite scripts or add results to a
system scrapbook. The scrapbooking is currently implemented + script
favorites to follow shortly.

mfitzp committed:
Add favorite support to scripts.

mfitzp committed:
Django compat additions for GenericForeignKey; slug url regex fix

mfitzp committed:
Re-arrangment of django_compat and settings.configure()

Using GenericForeignKeys requires the database configuration
to be up and running before creating models. This re-arranges
the django_compat file to import and `.configure()` settings
before performing other tasks.

Should now pass all tests.

mfitzp committed:
Fix some url/json errors in tests: json script/task views are still broken

Update tests with all_queues_json

Replace the not found task_error with 404

Use all_queues_json in tests

Further testing fixes

Further testing fixes

Add separate test for non-existent task; raise 404

mfitzp commented:
@Chris7 Working through fixing outstanding failing tests before polishing off the UI and have hit someone I can't figure out - I have added a `size_bytes` field to the `WooeyFile` class. There is an added migration to add this to an existing database.

When I run the test I get a `OperationalError: no such column: wooey_wooeyfile.size_bytes` but the database creation (+ the migration) both pass without problems. Any ideas?

mfitzp committed:
Add separate test for non-existent task; raise 404

mfitzp commented:
The failing tests are now fixed on Travis but am still not able to run it locally due the above error. Strange.

mfitzp committed:
Add metadata + lightbox to scrapbook

Chris7 commented:
Have you migrated the database? How are you running the tests locally? With
tox?

Sent from my phone
On Jul 25, 2015 2:15 PM, "Martin Fitzpatrick" notifications@github.com
wrote:

> The failing tests are now fixed on Travis but am still not able to run it
> locally due the above error. Strange.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/wooey/django-djangui/pull/16#issuecomment-124868909.

mfitzp commented:
I'm using `make test` to run the tests (which runs `nostests`) and this appears to correctly setup the database and apply the migrations. The specific migration `wooey.0004_wooeyfile_size_bytes` to add the column is also displayed during the setup.

Looking again however its actually the `coverage` run that fails. It says it creates the test database (`for alias 'default'`) but doesn't appear to apply the migrations.

Chris7 commented:
make test works fine for me. There were issues in getting it setup though. Here's some problems I saw that may be your cause:
1) .pyc files leftover from my branch upon checking out your branch. Make sure you don't have any lingering .pyc files in your migrations folder.
2) Your migration isn't actually committed to git. It shouldn't technically matter since the test will create it, but it'll also create auto-migrations for you. Check to make sure these are cleaned up.

And one last bit -- what Django/coveralls version are you using? I'm on Django 1.8.2 and coveralls 1.0a2. Are you using a virtualenv for this?

Chris7 commented on wooey/views/views.py:
```diff
@@ -65,23 +76,78 @@ def post(self, request, *args, **kwargs):
+
+        # Get the id of every favorite (scrapbook) file
+        ctype = ContentType.objects.get_for_model(Script)
+        ctx['favorite_script_ids'] = Favorite.objects.filter(content_type=ctype, user=self.request.user).values_list('object_id', flat=True)
```
This will fail if the user is Anonymous. We need to add a utility function for fetching users that will 1) evaluate self.request.user to make sure it isn't Lazy, 2) check if we are permitting anonymous users, 3) return an appropriate user for the anonymous case.

Chris7 commented on wooey/django_compat.py:
```diff
@@ -16,21 +22,23 @@ def __init__(self, content, mimetype='application/json', status=None, content_ty
 if DJANGO_VERSION >= DJ18:
+
     from django.template import Engine
+    from django.contrib.contenttypes.fields import GenericForeignKey
```
Can you move this to a separate place? I'd like to keep each segment specific to a given task -- like above this is 'Compatibilty for JSON', this was just 'compatibility for Template Rendering', etc.

mfitzp committed:
Re-structure django_compat to keep distinct sections

mfitzp commented:
Sure no problem, makes sense.

Chris7 commented on wooey/views/views.py:
```diff
@@ -65,23 +76,78 @@ def post(self, request, *args, **kwargs):
+
+
+        # Get the id of every favorite (scrapbook) file
+        ctype = ContentType.objects.get_for_model(Script)
```
We should initialize a dict of content types in our AppConfig. They're static so it's an easy place to avoid an extra db query.

Chris7 commented:
This looks awesome so far. It looks like its close to being finished so I threw in a few comments here and there. If you still have a bit to go, let me know so I don't bug you with optimizations you aren't doing yet because you're still hacking.

mfitzp commented:
Thanks it's fine, comment away! It is very close to being finished. Things still on my mental to-do list are -
- searching on the front page to filter
- user dashboard view (when user clicks on themselves) with some additional information, shortcuts to favorite scripts, latest outputs, etc.
- small bug where the scrapbook number doesn't update until you add/remove

Both of those shouldn't take more than a day or two to complete (most of the code is already in there) and then we'll be good to go. If you spot anything else obviously missing in the meantime let me know.

As I mentioned on the other PR #11 I think the best plan is probably to merge that then I can pull and re-base over it. I'll fix the conflicts if any and then merge this in to master. 

The actual merge into the old Wooey repo will take a bit of git foo so I'm thinking to create a dummy repo for testing the process + ensuring all the history is migrating correctly.

Chris7 commented:
Alright. Sounds good. I'm writing tests right now so I should be able to
move that branch in today or tomorrow.

Sent from my phone

Chris7 commented:
multiInput is merged.

mfitzp commented:
Awesome :+1: I'll rebase over that tomorrow and try and get the final bits of the UI finished off in the next couple of days

mfitzp commented:
Rebased on master; now to fix the test failures :)

Chris7 commented on wooey/models/favorite.py:
```diff
@@ -0,0 +1,18 @@
+
+    content_type = models.ForeignKey(ContentType)
+    object_id = models.PositiveIntegerField()
+    content_object = GenericForeignKey('content_type', 'object_id')
```
Which things are going to be favorited? I only ask because I sort of hate GFKs.

Consider this: You are fetching a user's favorites. There are 100 things. Now to get the thing that is favorited, you have an additional 2 queries per favorite (content table lookup, then model lookup). You can avoid this by fetching all favorites/gfks/etc. and making a huge dictionary for lookups -- but that can take up a ton of memory and cause the server to crash.

Do you know a good way around this? Maybe a custom manager can handle this because prefetches as far as I remember break down with GFKs as well.

mfitzp replied:
Just for Scripts and Files I imagine (can't think of anything else that would make sense). The scrapbook and the script favourites are the same thing.

You're right about the convoluted way to get the objects. At the moment the list of `Favorite` objects is got matching the right content type (id/etc.) and used to generate a list of object IDs. This list of IDs is then queried against the database as a list of IDs using `__in` (so 2 queries; plus another for the ContentType if it's not in memory).

It's been a while but you're right that with GFKs you lose prefetches (there is no real link in the ORM). The way around it I've used before is to use an adapter table, e.g. `ScriptFavorites` which would link from the `Script` to `Favorite`. This contains just the `script_id` and the `favorite_id` — the favorite info remains in a `Favorite` object.

The difference is that this allows you to traverse the ORM. The downside is you need a separate table for each relationship + the front end (JS) will be less generic (at the moment the content-type is used to route favorite changes).

Chris7 replied:
I'm good with the 2 query method. I forgot that approach. I think a custom manager would work best for that to assign the model attribute to the Favorite model after returning the query. So we can just work with something like:

[i.favorite for i in Favorite.objects.filter(user=user)]

and be able to forget any of the behind the scenes.

mfitzp committed:
Implement AJAX script search from home

Fix tests

mfitzp committed:
Add combined migrations for UI changes; .gitignore tests media folder.

mfitzp committed:
Remove migration dependency on contenttypes name column deletion, does not affect our code.

mfitzp committed:
Add simple job list view with limit on dropdown and link

mfitzp committed:
Fix tests for new JSON structure

mfitzp commented:
@Chris7 So I think we're nearly there, a bit later than expected — various bits of work cropped up over the past week — but better late than never.

It isn't _finished_ but I think the bulk of the work is done and it's now a case of tweaking, template designing, bug fixing, etc. which may be better handled in master/smaller branches?

One thing that I think might be worth a look —
- job vs. task which are used interchangeably (unless there is a distinction I'm missing)
- some models have redundant named fields, eg. script.script_name vs. script.name

...but that might be better handled in another branch?

Chris7 commented:
@mfitzp 
There is a distinction between the two...at least there should be. A job is the model which stores a given run whereas a task is the execution of the job.

For the redundant fields, I'm ok with simplifying it, though I'll provide my rationale. You can assign a field that is a method and cause yourself a lot of unneeded grief. I spent like 4 hours figuring out a bug that ended up being that situation where I overrode an undocumented model method with a field name.

I'm good with merging into master so we can both work on something stable. I've been sitting back for a little because there are a lot of changes that are coming in.

mfitzp commented:
Cool. I can see the distinction in the models (WooeyJob vs. CeleryTask) the question was more about templates/variables/urls. The output of a job is displayed in a template called `task/task_view` with a variable called `task_info` which contains job info. I'm wondering if it would make sense to be stricter about this and use the terminology 'job' for all the WooeyJob related things?

Ha I have no desire to waste 4 hours tracking something like that down - the field names stay as is. We can revisit this at some point in the future if we feel like it.

Yep, would be good to both be able to pick off the small bits and polish up.

Chris7 commented:
Ah, I see what you mean and I agree for using 'job' because 'task' is really specific to Celery things. I think the mix-up came about by providing information about the task's progress and just sticking things in there without too much thought.

mfitzp commented:
It probably made a lot more sense before I started butchering it for the UI changes. 

I'll make the task->job changes this evening/tomorrow and then this branch should be good to merge.

mfitzp committed:
Standardise on using job terminology for front-end.

Renamed various things so that the front-end refers to jobs
only. Back-end interaction with celery is unchanged. Should
make the distinction clearer.

mfitzp commented:
Sorted out the job/task terminology on the front end. Merging this now so we can hack away in master.

Chris7 commented:
It looks awesome. Great job -- especially on the task view.
